src/Common/src/System/Security/Cryptography/AsnReader.cs

```diff
@@ -0,0 +1,2802 @@
+                const ushort RedundancyMask = 0b1111_1111_1000_0000;
+                ushort masked = (ushort)(bigEndianValue & RedundancyMask);
+
+                // If the first 9 bits are all 0 or are all 1, the value is invalid.
+                if (masked == 0 || masked == RedundancyMask)
+                {
+                    throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
+                }
+            }
+
+            return contents;
+        }
+
+        public ReadOnlyMemory<byte> GetIntegerBytes() =>
+            GetIntegerBytes(Asn1Tag.Integer);
+
+        public ReadOnlyMemory<byte> GetIntegerBytes(Asn1Tag expectedTag)
+        {
+            ReadOnlyMemory<byte> contents =
+                GetIntegerContents(expectedTag, UniversalTagNumber.Integer, out int headerLength);
+
+            _data = _data.Slice(headerLength + contents.Length);
+            return contents;
+        }
+
+        private bool TryReadSignedInteger(
+            int sizeLimit,
+            Asn1Tag expectedTag,
+            UniversalTagNumber tagNumber,
+            out long value)
+        {
+            Debug.Assert(sizeLimit <= sizeof(long));
+
+            ReadOnlyMemory<byte> contents = GetIntegerContents(expectedTag, tagNumber, out int headerLength);
+
+            if (contents.Length > sizeLimit)
+            {
+                value = 0;
+                return false;
+            }
+
+            ReadOnlySpan<byte> contentSpan = contents.Span;
+
+            bool isNegative = (contentSpan[0] & 0x80) != 0;
+            long accum = isNegative ? -1 : 0;
+
+            for (int i = 0; i < contents.Length; i++)
+            {
+                accum <<= 8;
+                accum |= contentSpan[i];
+            }
+
+            _data = _data.Slice(headerLength + contents.Length);
+            value = accum;
+            return true;
+        }
+
+        private bool TryReadUnsignedInteger(
+            int sizeLimit,
+            Asn1Tag expectedTag,
+            UniversalTagNumber tagNumber,
+            out ulong value)
+        {
+            Debug.Assert(sizeLimit <= sizeof(ulong));
+
+            ReadOnlyMemory<byte> contents = GetIntegerContents(expectedTag, tagNumber, out int headerLength);
+            ReadOnlySpan<byte> contentSpan = contents.Span;
+            int contentLength = contents.Length;
+
+            bool isNegative = (contentSpan[0] & 0x80) != 0;
+
+            if (isNegative)
+            {
+                value = 0;
+                return false;
+            }
+
+            // Ignore any padding zeros.
+            if (contentSpan.Length > 1 && contentSpan[0] == 0)
+            {
+                contentSpan = contentSpan.Slice(1);
+            }
+
+            if (contentSpan.Length > sizeLimit)
+            {
+                value = 0;
+                return false;
+            }
+
+            ulong accum = 0;
+
+            for (int i = 0; i < contentSpan.Length; i++)
```

Comment:
Would it make sense to use BinaryPrimitives?https://github.com/dotnet/corefx/blob/master/src/System.Memory/ref/System.Memory.cs#L232

Reply:
BinaryPrimitives looks like it requires the full value-length to do a read.  So TryReadUInt64 will fail if 8 bytes aren't available.  DER rules trim superfluous bytes (since they're length prepended), so 0x007FFFFF_FFFFFFFF will only have 7 bytes recorded, making it not something that can be read.

Reply:
Interesting. If we see this more often, maybe we should add some overloads to BinaryPrimitives to handle that.

cc: @ahsonkhan